File: Flink0830Tutorial/src/main/scala/com/atguigu/course/KeyedProcessFunctionExample.scala
package com.atguigu.course

import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.api.scala.typeutils.Types
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

// 教程中例子6-5
// 检测连续 1s 温度上升，然后报警
object KeyedProcessFunctionExample {

  // KeyedProcessFunction只能操作KeyedStream
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    val warnings = env
      .addSource(new SensorSource)
      .keyBy(_.id)
      // 注意是 process ！
      .process(new TempIncreaseAlertFunction)

    warnings.print()

    env.execute()
  }

  class TempIncreaseAlertFunction extends KeyedProcessFunction[String, SensorReading, String] {
    // 保存上一个传感器温度值
    // 惰性赋值
    // 只会初始化一次，当程序挂掉，再重启的时候，会调用getState方法看一下这个状态变量存在不存在，如果存在，就不初始化
    // 默认值为 0.0
    // 可以把状态变量存储到状态后端（JVM，HDFS，RocksDB，FileSystem）
    // 供程序挂掉恢复时，从检查点恢复使用
    lazy val lastTemp : ValueState[Double] = getRuntimeContext.getState(
      new ValueStateDescriptor[Double]("lastTemp", Types.of[Double])
    )

    // 默认值为 0L
    // 保存定时器时间戳的状态变量
    lazy val currentTimer : ValueState[Long] = getRuntimeContext.getState(
      new ValueStateDescriptor[Long]("timer", Types.of[Long])
    )

    // 每到来一条事件，都会调用一次
    override def processElement(value: SensorReading,
                                // #是类型投影的意思，用来访问内部类
                                ctx: KeyedProcessFunction[String, SensorReading, String]#Context,
                                out: Collector[String]): Unit = {
      // ValueState的读取使用`.value()`方法
      // 如果当前读数是传感器的第一条数据，那么 prevTemp 的值是 0.0
      val prevTemp = lastTemp.value()

      // ValueState的更新使用`.update()`方法
      lastTemp.update(value.temperature)

      val curTimerTimestamp = currentTimer.value()

      if (prevTemp == 0.0 || value.temperature < prevTemp) {
        // 温度下降，或者value是第一个温度读数，删除定时器
        ctx.timerService().deleteProcessingTimeTimer(curTimerTimestamp)
        // ValueState清空操作使用`.clear()`
        currentTimer.clear()
      } else if (value.temperature > prevTemp && curTimerTimestamp == 0) {
        // 当前机器时间 1s 之后的时间戳
        val timerTs = ctx.timerService().currentProcessingTime() + 1000
        // 在时间戳 timerTs 注册一个定时事件
        ctx.timerService().registerProcessingTimeTimer(timerTs)
        // 将时间戳存储到状态变量
        currentTimer.update(timerTs)
      }
    }

    // onTimer 向下发送数据
    override def onTimer(timestamp: Long,
                         ctx: KeyedProcessFunction[String, SensorReading, String]#OnTimerContext,
                         out: Collector[String]): Unit = {
      out.collect("传感器id为： " + ctx.getCurrentKey + "的传感器温度值已经连续 1s 上升了。")
      // 别忘了清空状态变量
      currentTimer.clear()
    }

  }

}

File: flink-code-java-0421/src/main/java/com/atguigu/day03/KeyedProcessFunctionExample.java
package com.atguigu.day03;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

// KeyedProcessFunction在分流以后调用
// ProcessWindowFunction和AggregateFunction在分流、开窗以后调用
public class KeyedProcessFunctionExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<String> stream = env.socketTextStream("localhost", 9999);

        stream
                .flatMap(new FlatMapFunction<String, Tuple2<String, String>>() {
                    @Override
                    public void flatMap(String s, Collector<Tuple2<String, String>> collector) throws Exception {
                        String[] arr = s.split(" ");
                        collector.collect(Tuple2.of(arr[0], arr[1]));
                    }
                })
                .keyBy(r -> r.f0)
                .process(new MyKeyed())
                .print();

        env.execute();
    }

    public static class MyKeyed extends KeyedProcessFunction<String, Tuple2<String, String>, String> {
        // 每来一条数据，调用一次
        @Override
        public void processElement(Tuple2<String, String> stringStringTuple2, Context context, Collector<String> collector) throws Exception {
            long tenSencondLater = context.timerService().currentProcessingTime() + 10 * 1000L;
            context.timerService().registerProcessingTimeTimer(tenSencondLater);
        }

        @Override
        public void onTimer(long timestamp, OnTimerContext ctx, Collector<String> out) throws Exception {
            super.onTimer(timestamp, ctx, out);
            out.collect("定时器触发了！");
        }
    }
}
